src/passes/PostAssemblyScript.cpp

```diff
@@ -0,0 +1,530 @@
+    localSet->value = call->operands[0];
+  }
+
+  void eliminateRelease(Expression** location) {
+    // assumes __release(local.get(X, ...))
+    if (auto* call = (*location)->dynCast<Call>()) {
+      assert(call->target == RELEASE);
+      assert(call->operands.size() == 1);
+      assert(call->operands[0]->is<LocalGet>());
+      Builder builder(*getModule());
+      *location = builder.makeNop();
+    } else {
+      assert((*location)->is<Nop>()); // already replaced
+    }
+  }
+
+  // Tests if a retain reaches an escape and thus is
+  // considered necessary.
+  bool testReachesEscape(LocalSet* retain, AliasGraph& graph) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      if (releases.find(localGet) != releases.end()) {
+        continue;
+      }
+      if (escapes.find(localGet) != escapes.end()) {
+        return true;
+      }
+      for (auto* localSet : graph.getInfluences[localGet]) {
+        if (testReachesEscape(localSet, graph)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  // Collects all reachable releases of a retain
+  void collectReleases(LocalSet* retain,
+                       AliasGraph& graph,
+                       std::unordered_set<Expression**>& found) {
+    for (auto* localGet : graph.setInfluences[retain]) {
+      auto foundRelease = releases.find(localGet);
+      if (foundRelease != releases.end()) {
+        found.insert(foundRelease->second);
+      } else {
+        for (auto* localSet : graph.getInfluences[localGet]) {
+          collectReleases(localSet, graph, found);
+        }
+      }
+    }
+  }
+
+  // Given a retain, gets the retained expression
+  static Expression* getRetainedExpression(LocalSet* retain) {
+    // assumes local.set(X, __retain(...))
+    auto* retainCall = retain->value->cast<Call>();
+    assert(retainCall->operands.size() == 1);
+    return retainCall->operands[0];
+  }
+
+  // Tests if a retained value originates at an allocation
+  // and thus is considered necessary.
+  bool testRetainsAllocation(Expression* retained, AliasGraph& graph) {
+    if (auto* call = retained->dynCast<Call>()) {
+      if (call->target == ALLOC || call->target == ALLOCARRAY) {
+        return true;
+      }
+    } else {
+      if (auto* localGet = retained->dynCast<LocalGet>()) {
+        for (auto* localSet : graph.getSetses[localGet]) {
+          if (localSet != nullptr &&
+              testRetainsAllocation(localSet->value, graph)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  // Given a release location, gets the release
+  static LocalGet* getReleaseByLocation(Expression** releaseLocation) {
+    // assumes __release(local.get(X, ...))
+    auto* releaseCall = (*releaseLocation)->cast<Call>();
+    assert(releaseCall->operands.size() == 1);
+    return releaseCall->operands[0]->cast<LocalGet>();
+  }
+
+  // Tests if a release has balanced retains
```

Comment:
Would be good to define what "balanced" means. Not sure if I missed that somewhere earlier.

Reply:
Expanded the comment. Basically means that all branches (or paths through AliasGraph) are coming from a retain or leading to a release etc. I pretty much made that terminology up :)